### ISSUE TITLE
Running iperender in parallel doesn't work because tempfiles are deterministic

### ISSUE THREAD

**Issue by jyn514** [otfried/ipe#375]
```
> iperender -pdf nontextbook_diagram.ipe nontextbook_diagram.pdf
Pdflatex was run sucessfully.
> iperender -pdf 1.43_diagram.ipe 1.43_diagram.pdf
Pdflatex was run sucessfully.
> iperender -pdf nontextbook_diagram.ipe nontextbook_diagram.pdf & iperender -pdf 1.43_diagram.ipe 1.43_diagram.pdf
[1] 13659
There was an error reading the Pdflatex output.
Pdflatex was run sucessfully.
[1]+  Exit 1                  iperender -pdf nontextbook_diagram.ipe nontextbook_diagram.pdf
> strace -e openat 2>&1 iperender -pdf 1.43_diagram.ipe 1.43_diagram.pdf | grep -v RDONLY
openat(AT_FDCWD, "/home/joshua/.ipe/latexrun/ipetemp.tex", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=13841, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
Pdflatex was run sucessfully.
openat(AT_FDCWD, "1.43_diagram.pdf", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 3
+++ exited with 0 +++
```
IPE should use a different temp file for each invocation, either using `tmpfile()` or the PID of the current process.

**Comment by jyn514** [comment]
I'm willing to work on a patch for this, but I can't seem to find the source code. https://github.com/otfried/ipe only has a readme.

**Comment by charlesdaniels** [comment]
@jyn514 source code is located here http://ipe.otfried.org/

The relevant portions of the code are in `src/ipelib/ipedoc.cpp`, line 633, and in `src/ipelib/ipeplatform.cpp`, line 488 (used two places in the function body).

I would suggest creating an `ipetempfilename()` method somewhere and using it in both places. It looks like in `ipedoc.cpp`, the same basename is used for the log and PDF file, so maybe that method just returns the basename, or accepts the extension as an argument.

**Comment by jyn514** [comment]
This fixed the issue:
```diff
diff --git a/include/ipebase.h b/include/ipebase.h
index bf41e26..aa0aea2 100644
--- a/include/ipebase.h
+++ b/include/ipebase.h
@@ -409,5 +409,7 @@ namespace ipe {
 
 } // namespace
 
+extern ipe::String ipetempfile() noexcept;
+
 // --------------------------------------------------------------------
 #endif
diff --git a/ipelib/ipedoc.cpp b/ipelib/ipedoc.cpp
index 28351cc..eb48a04 100644
--- a/ipelib/ipedoc.cpp
+++ b/ipelib/ipedoc.cpp
@@ -630,9 +630,10 @@ int Document::runLatexAsync(String docname, String &texLog, Latex **pConverter)
   if (latexDir.empty())
     return ErrNoDir;
 
-  String texFile = latexDir + "ipetemp.tex";
-  String pdfFile = latexDir + "ipetemp.pdf";
-  String logFile = latexDir + "ipetemp.log";
+  String basename = ipetempfile();
+  String texFile = latexDir + basename + ".tex";
+  String pdfFile = latexDir + basename + ".pdf";
+  String logFile = latexDir + basename + ".log";
 
   std::remove(logFile.z());
 
diff --git a/ipelib/ipeplatform.cpp b/ipelib/ipeplatform.cpp
index 66cb3f6..57e5483 100644
--- a/ipelib/ipeplatform.cpp
+++ b/ipelib/ipeplatform.cpp
@@ -231,6 +231,10 @@ void Platform::initLib(int version)
 #endif
 }
 
+String ipetempfile() noexcept {
+    return String("ipetemp") + std::to_string(getpid()).c_str();
+}
+
 //! Enable or disable display of ipeDebug messages.
 void Platform::setDebug(bool debug)
 {
@@ -501,6 +505,7 @@ int Platform::runLatex(String dir, LatexType engine, String docname) noexcept
     if (i > 0)
       texinputs = docname.substr(0, i-1);
   }
+  String basename = ipetempfile();
 #ifdef WIN32
   if (!online && getenv("IPETEXFORMAT")) {
     latex = (engine == LatexType::Xetex) ?
@@ -536,7 +541,7 @@ int Platform::runLatex(String dir, LatexType engine, String docname) noexcept
       bat += ";%PATH%\r\n";
     }
     bat += latex;
-    bat += " ipetemp.tex\r\n";
+    bat += " " + basename + ".tex\r\n";
   }
   if (!texinputs.empty())
     bat += "endlocal\r\n";
@@ -593,7 +598,7 @@ int Platform::runLatex(String dir, LatexType engine, String docname) noexcept
   }
   String s("cd \"");
   s += dir;
-  s += "\"; rm -f ipetemp.log; ";
+  s += String("\"; rm -f ") + basename + ".log; ";
   if (!texinputs.empty()) {
     s += "export TEXINPUTS=\"";
     s += texinputs;
@@ -614,7 +619,7 @@ int Platform::runLatex(String dir, LatexType engine, String docname) noexcept
       s += latex;
     else
       s += String("\"") + path + "/" + latex + "\"";
-    s += " ipetemp.tex";
+    s += String(" ") + basename + ".tex";
   }
   s += " > /dev/null";
   int result = std::system(s.z());
```

**Comment by otfried** [comment]
What's the motivation for running iperender in parallel?   If you replace the `&` by `&&` it would work...

I very often have to help Ipe users debug Latex problems, and your patch would make that much harder.

**Comment by otfried** [comment]
If you really need to run iperender in parallel (I assume this happens in a parallel make), I suggest you run the following parallel-safe shell script to execute iperender:
```
#!/bin/bash
dir=`mktemp -d`
export IPELATEXDIR=$dir
iperender $*
rm -fr $dir
```

**Comment by jyn514** [comment]
> I very often have to help Ipe users debug Latex problems, and your patch would make that much harder.

What if there were an opt-in way to make the tempfile deterministic (or to pass an explicit path for the tempfile)? Then you would still be able to debug problems, but running in parallel would work by default.

> If you really need to run iperender in parallel (I assume this happens in a parallel make), I suggest you run the following parallel-safe shell script to execute iperender

Thanks, that does fix the issue. It would be nice to have it work by default though, since I'm using the same makefile with several people in a project (you assume correctly) and they'd all have to set up the wrapper.
